### PR TITLE
Specify fields for get/list commands

### LINKED ISSUE #4: Specify table keys for GET & LIST commands
Although our get/list commands display sensible keys by default, it would be helpful to specify what information to output. This should accept all properties listed on our [API schemas](https://api.spinupwp.com/#schemas).

`spinupwp servers:list --fields=id,ip_address,ssh_port,ubuntu_version`

The first time a user supplies the `--fields` option, we should prompt them whether they want to save these fields as the default for the current command. If they select 'y', this should be saved to their current profile and would overwrite our sensible defaults.

### PR DESCRIPTION
**Pull request by danielmlozano**
Resolves #4 

You can pass `--fields=<fields,list>` and it will ask if you want to save the fields as default for the current command.

### CODE REVIEW

**Comment by katelynnbarlowe** [comment]
@danielmlozano Just a reminder this is good for moving forward since the SSH commands are now merged.

**Comment by A5hleyRich** [comment]
@danielmlozano I'm not sure about the `--savefields` flag. If we're going to allow the defaults to be reconfigured, I think we should add them to the configure command, maybe under a flag like `spinupwp configure --fields`. But for now, let's not provide that functionality.

The first time a user supplies the `--fields` option, we should prompt them whether they want to save these fields as the default for the current command. If they select 'y', this should be saved to their current profile and would overwrite our sensible defaults. If they wish to change them in the future, they'll need to modify their config.json file.

**Comment by danielmlozano** [comment]
@A5hleyRich Yup. the first time you run the command it will prompt whether you want to save the given fields as default. If "yes", the fields are saved in the current profile, otherwise the fields configuration for the given command will be saved empty so we don't ask again, hence the "`--savefields`" flag, if your answer is "no" the first time, you should be able to save the defaults fields later, or even if you save your own default fields, you may want to change them in the future. Should we do it in the `configure` command then?

**Comment by A5hleyRich** [comment]
@danielmlozano as there's a workaround, let's keep this simple and not allow them to update the default columns once prompted the first time. We'll update the README with possible configs and where to change them.

> If they wish to change them in the future, they'll need to modify their config.json file.

**Comment by danielmlozano** [comment]
@katelynnbarlowe this is ready for your review.

File: app/Helpers/Configuration.php [hunk 1]
```diff
@@ -53,6 +53,28 @@ public function set(string $key, string $value, string $profile = 'default'): vo
...
+        $config = $this->config;
+
+        $command = trim($command);
+
+        Arr::set($config, "{$profile}.command_options.{$command}.{$key}", $value);
+
+        file_put_contents($this->configFilePath(), json_encode($config, JSON_PRETTY_PRINT));
+
+        $this->config = $config;
+    }
```
Comment (A5hleyRich) [inline app/Helpers/Configuration.php:76]: There's some repetition here. Can we not call `get` and `set` from within these methods?

**Review by A5hleyRich — CHANGES_REQUESTED** [review]
@danielmlozano passing the `--fields` flag when defaults have already been saved wipes those defaults.

For sites, how would I specify a nested column? For example, what's the expected format to display the ID, domain, site user, and git branch?

**Comment by danielmlozano** [comment]
> For sites, how would I specify a nested column? For example, what's the expected format to display the ID, domain, site user, and git branch?

@A5hleyRich With the current version, you can use dot notation, so to display ID, domain, site user and git branch, the option will be: `--fields=id,domain,site_user,git.branch`

File: app/Commands/Concerns/SpecifyFields.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    protected function saveFieldsFilter($saveConfiguration = false): void
+    {
+        $commandOptions = $this->config->getCommandConfiguration($this->command, $this->profile());
+
+        if (!empty($commandOptions)) {
+            return;
+        }
+
+        if (empty($commandOptions) && !$saveConfiguration) {
+            $saveConfiguration = $this->confirm('Do you want to save the specified fields as default for this command?', true);
```
Comment (A5hleyRich) [inline app/Commands/Concerns/SpecifyFields.php:88]: > Do you want to save the specified fields as the default for this command?

**Review by A5hleyRich — CHANGES_REQUESTED** [review]
Nice work @danielmlozano. This could do with a run of phpstan.

To clarify expected output:

- [x] No command default or `--fields` flag
  - [x] Table: Sensible defaults (as before)
  - [x] JSON: All data (as before)
- [x] Command default set or `--fields` flag supplied
  - [x] Table: Only fields data
  - [x] JSON: Only fields data

**Review by Jamesclark32 — CHANGES_REQUESTED** [review]
Nice work! This feels like a really solid solution to a complex problem. I'm bad at using closures, and this PR is inspiring me - you've solved some really complex problems very cleanly by using them for filters and ignore rules. 

The `SpecifyFields` class does it's job well, which suggests the underlying solution is a good approach. However, the `SpecifyFields` class has to do a lot of heavy lifting to fully process the `fieldMaps`. 

I think fleshing out the data structure for `fieldMaps` would build on your ideas and take them to the next level. I also think there is a pattern here much like Ryan has been working against in https://github.com/deliciousbrains/spinupwp-cli/pull/28 . What I'm proposing below is a similar solution to the path he is on there. 

If we built out a generic `Field` class and instantiated a series of those in each `Command` class in place of the array items, we could __really__ streamline things. Most of the complex things you've had to do in `SpecifyFields` could be refactored into that `Fields` class.

A roughed in proof of concept for a `Field` class:
```php
class Field
{
    public string $name;
    public array $aliases = [];
    public string $displayName;
    public ?Closure $shouldIgnoreRule = null;
    public ?Closure $transformRule = null;

    public function __construct(string $displayName)
    {
        $this->displayName = $displayName;
    }

    public function addAlias(string $alias): self
    {
        $this->aliases[] = $alias;
        return $this;
    }

    public function withTransformRule(Closure $filter): self
    {
        $this->transformRule = $filter;
        return $this;
    }

    public function withIgnoreRule(Closure $filter): self
    {
        $this->transformRule = $filter;
        return $this;
    }

  public function getDisplayLabel(bool $isLongVersion = false): string 
  { 
       if ($isLongVersion) { 
          return $this->displayName;
       }
       return $this->name;
  }

    public function getDisplayValue($resource)
    {
        if (!$this->transformRule) {
            return $resource->{$this->name};
        }

        return $this->filterRule($resource->{$this->name});
    }

    public function shouldIgnore($resource): bool
    {
        if (!$this->shouldIgnoreRule) {
            return false;
        }

        return $this->shouldIgnoreRule($resource->{$this->name});
    }

    public function shouldDisplay($resource): bool
    {
        return !$this->shouldIgnore($resource);
    }

    public function isInFilter(array $filterFields): bool
    {
        if (in_array($this->name, $filterFields)) {
            return true;
        }

        $common = array_intersect($this->aliases, $filterFields);
        return count($common) > 0;
    }
}
```
With the fluent methods, building them out is pretty clean.
```php
  'Database Backups' => [
                'property' => 'backups|backups.database',
                'filter'   => fn ($value)   => $value['database'] ? 'Enabled' : 'Disabled',
            ],
```
becomes: 
```php
(new Field('Database Backups'))
    ->withName('backups')
    ->withAlias('backups.database')
    ->withTransformRule(fn($value) => $value['database'] ? 'Enabled' : 'Disabled');
```

and we have a number of methods ready to make consuming them like we do in `SpecifiyFields` much cleaner.  For example, 

```php
$fields[$field->getDisplayLabel($this->displayFormat() === 'table')] = $field->getDisplayValue($resource);
```
 Replaces  [these calculations](https://github.com/deliciousbrains/spinupwp-cli/pull/17/files#diff-24b84cdf23b2543a879fa16a6dedea64454c5e07c3f9ec288859f510d48c32b9R44) 
 
 I'm not sure how best to break this all down into a checkboxable list here.... 
  
 - [ ] Refactor `Fields` concept into a class. 
 - [ ] Update commands to consume the `Fields` classes.
 - [ ] Update SpecificFields to use `Fields` classes.

File: app/Commands/Concerns/SpecifyFields.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            return $resource->toArray();
+        }
+
+        $fields = [];
+
+        $commandFields = $this->config->getCommandConfiguration($this->command, $this->profile())['fields'] ?? null;
+
+        if ($commandFields) {
+            $fieldsFilter = explode(',', str_replace(' ', '', $commandFields));
+        }
```
[the remark this answers is not in the document]
Comment (danielmlozano) [reply app/Commands/Concerns/SpecifyFields.php]: ✅

File: app/Commands/Concerns/SpecifyFields.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+        if ($this->option('fields')) {
+            $fieldsOption = str_replace(' ', '', strval($this->option('fields')));
```
[the remark this answers is not in the document]
Comment (danielmlozano) [reply app/Commands/Concerns/SpecifyFields.php:25]: If we don't use strval, Phpstan will complain about types.

File: app/Commands/Concerns/SpecifyFields.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+        if ($saveConfiguration) {
+            $this->config->setCommandConfiguration($this->command, 'fields', strval($this->option('fields')), $this->profile());
+            return;
+        }
+
+        $this->config->setCommandConfiguration($this->command, 'fields', '', $this->profile());
+    }
+
+    protected function applyFilter(array $fieldsFilter): void
```
[the remark this answers is not in the document]
Comment (danielmlozano) [reply app/Commands/Concerns/SpecifyFields.php]: ✅

File: app/Commands/Concerns/SpecifyFields.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                $fields[$this->displayFormat() === 'table' ? $name : $resourceProp['property']] = $value;
+                continue;
+            }
+            $fields[$this->displayFormat() === 'table' ? $name : $resourceProp] = $resource->{$resourceProp};
+        }
+
+        return $fields;
+    }
+
+    protected function propertyInFilter(string $property, array $fieldsFilter): bool
```
[the remark this answers is not in the document]
Comment (danielmlozano) [reply app/Commands/Concerns/SpecifyFields.php]: ✅

**Review by Jamesclark32 — CHANGES_REQUESTED** [review]
I really like how this looks with the `Fields` class! 

Does the `--savefields` parameter still needs to be implemented? 

There was a lot of discussion about how best to make reusable parameter definitions in another PR, which lead to this issue: https://github.com/deliciousbrains/spinupwp-cli/issues/32 

https://github.com/deliciousbrains/spinupwp-cli/pull/28#issuecomment-1025930697 had some relevant context, too. 

The pattern being discussed *might* be useful for this need? I'm not sure how easily the concern of adding that parameter can land in the SpecifyFeilds trait without complicating interaction with the InteractsWithIo trait. Potentially a single trait could be responsible for adding the `--savefields` and `--fields parameters`. 

What are your thoughts on that approach, both in general and specific to this need?

- [x] Missing return type hint on `App/Fields::getDisplayValue()`
- [ ] Add `--savefields` functionality

**Comment by danielmlozano** [comment]
@Jamesclark32 the `--savefields` flag is not longer necessary.

**Review by Jamesclark32 — APPROVED** [review]
This PR is looking great! 🥳

**Pull request merged**